runtime/near-vm-runner/src/prepare.rs

```diff
@@ -172,26 +197,175 @@ impl<'a> ContractModule<'a> {
+        /// we reject such a module.
+        fn ensure_no_internal_memory(self) -> Result<Self, PrepareError> {
+            if self.module.memory_section().map_or(false, |ms| !ms.entries().is_empty()) {
+                Err(PrepareError::InternalMemoryDeclared)
+            } else {
+                Ok(self)
+            }
+        }
+
+        fn inject_gas_metering(self) -> Result<Self, PrepareError> {
+            let Self { module, config } = self;
+            // Free config, no need for gas metering.
+            if config.regular_op_cost == 0 {
+                return Ok(Self { module, config });
+            }
+            let gas_rules = pwasm_utils::rules::Set::new(1, Default::default())
+                .with_grow_cost(config.grow_mem_cost);
+            let module = pwasm_utils::inject_gas_counter(module, &gas_rules)
+                .map_err(|_| PrepareError::GasInstrumentation)?;
+            Ok(Self { module, config })
+        }
+
+        fn inject_stack_height_metering(self) -> Result<Self, PrepareError> {
+            let Self { module, config } = self;
+            let module = pwasm_utils::stack_height::inject_limiter(
+                module,
+                config.limit_config.max_stack_height,
+            )
+            .map_err(|_| PrepareError::StackHeightInstrumentation)?;
+            Ok(Self { module, config })
+        }
+
+        /// Scan an import section if any.
+        ///
+        /// This accomplishes two tasks:
+        ///
+        /// - checks any imported function against defined host functions set, incl.
+        ///   their signatures.
+        /// - if there is a memory import, returns it's descriptor
+        fn scan_imports(self) -> Result<Self, PrepareError> {
+            let Self { module, config } = self;
+
+            let types = module.type_section().map(elements::TypeSection::types).unwrap_or(&[]);
+            let import_entries =
+                module.import_section().map(elements::ImportSection::entries).unwrap_or(&[]);
+
+            let mut imported_mem_type = None;
+
+            for import in import_entries {
+                if import.module() != "env" {
+                    // This import tries to import something from non-"env" module,
+                    // but all imports are located in "env" at the moment.
+                    return Err(PrepareError::Instantiate);
+                }
+
+                let type_idx = match *import.external() {
+                    External::Function(ref type_idx) => type_idx,
+                    External::Memory(ref memory_type) => {
+                        imported_mem_type = Some(memory_type);
+                        continue;
+                    }
+                    _ => continue,
+                };
+
+                let Type::Function(ref _func_ty) =
+                    types.get(*type_idx as usize).ok_or(PrepareError::Instantiate)?;
+
+                // TODO: Function type check with Env
```

Comment:
optional: I'm not a fan of 'orphaned' TODOs ;-) we should either have a github/jira issue that is tracking it (if this is a lot more complex fix), or author name + additional info on why this is complex enough that we're not fixing it in this PR.